// dialute/dist/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dialute = void 0;
const tslib_1 = require("tslib");
// @ts-ignore
const express = require("express");
const dialog_1 = require("./dialog");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
class Dialute {
    constructor({ dm, port = '8000' }) {
        this.dm = dm;
        this.app = express();
        this.port = port;
        this.app.use((req, res, next) => {
            (0, dialog_1.dateLog)(`${chalk_1.default.green(req.method)}: ${chalk_1.default.cyan(req.path)}`);
            next();
        });
        this.app.use(express.json());
        this.app.post('/app-connector/', (request, response) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const body = (yield this.dm.process(request.body)).body;
            response.send(body);
        }));
    }
    static fromEntrypoint(entrypoint) {
        const dm = new dialog_1.DialogManager(entrypoint);
        const port = '8000';
        return new Dialute({ dm, port });
    }
    shareApp(path) {
        this.app.use(express.static(path));
        return this;
    }
    start() {
        this.app.listen(this.port, () => (0, dialog_1.dateLog)(chalk_1.default.blue(`Start server on http://localhost:${this.port}/`)));
    }
}
exports.Dialute = Dialute;
